File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Benchmarks/Models/BenchmarkTypes.swift
//
//  BenchmarkTypes.swift
//  RunAnywhereAI
//
//  Data models for the benchmarking suite.
//

import Foundation
import RunAnywhere

// MARK: - Benchmark Category

enum BenchmarkCategory: String, CaseIterable, Codable, Sendable, Identifiable {
    case llm
    case stt
    case tts
    case vlm
    case diffusion

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .llm: return "LLM"
        case .stt: return "STT"
        case .tts: return "TTS"
        case .vlm: return "VLM"
        case .diffusion: return "Diffusion"
        }
    }

    var iconName: String {
        switch self {
        case .llm: return "text.bubble"
        case .stt: return "waveform"
        case .tts: return "speaker.wave.3"
        case .vlm: return "eye"
        case .diffusion: return "paintbrush"
        }
    }

    var modelCategory: ModelCategory {
        switch self {
        case .llm: return .language
        case .stt: return .speechRecognition
        case .tts: return .speechSynthesis
        case .vlm: return .multimodal
        case .diffusion: return .imageGeneration
        }
    }
}

// MARK: - Benchmark Run Status

enum BenchmarkRunStatus: String, Codable, Sendable {
    case running
    case completed
    case failed
    case cancelled
}

// MARK: - Benchmark Scenario

struct BenchmarkScenario: Codable, Sendable, Identifiable {
    let name: String
    let category: BenchmarkCategory
    let parameters: [String: String]?

    var id: String { "\(category.rawValue)_\(name)" }

    init(name: String, category: BenchmarkCategory, parameters: [String: String]? = nil) {
        self.name = name
        self.category = category
        self.parameters = parameters
    }
}

// MARK: - Component Model Info (snapshot of ModelInfo for persistence)

struct ComponentModelInfo: Codable, Sendable {
    let id: String
    let name: String
    let framework: String
    let category: String

    init(from model: ModelInfo) {
        self.id = model.id
        self.name = model.name
        self.framework = model.framework.displayName
        self.category = model.category.rawValue
    }
}

// MARK: - Device Info (snapshot for persistence)

struct BenchmarkDeviceInfo: Codable, Sendable {
    let modelName: String
    let chipName: String
    let totalMemoryBytes: Int64
    let availableMemoryBytes: Int64
    let osVersion: String

    static func fromSystem(_ info: SystemDeviceInfo) -> BenchmarkDeviceInfo {
        BenchmarkDeviceInfo(
            modelName: info.modelName,
            chipName: info.chipName,
            totalMemoryBytes: info.totalMemory,
            availableMemoryBytes: SyntheticInputGenerator.availableMemoryBytes(),
            osVersion: info.osVersion
        )
    }
}

// MARK: - Benchmark Metrics

struct BenchmarkMetrics: Codable, Sendable {
    // Common
    var endToEndLatencyMs: Double = 0
    var loadTimeMs: Double = 0
    var warmupTimeMs: Double = 0
    var memoryDeltaBytes: Int64 = 0

    // LLM-specific
    var ttftMs: Double?
    var tokensPerSecond: Double?
    var inputTokens: Int?
    var outputTokens: Int?

    // STT-specific
    var audioLengthSeconds: Double?
    var realTimeFactor: Double?

    // TTS-specific
    var audioDurationSeconds: Double?
    var charactersProcessed: Int?

    // VLM-specific
    var promptTokens: Int?
    var completionTokens: Int?

    // Diffusion-specific
    var generationTimeMs: Double?

    // Error info
    var errorMessage: String?
    var didSucceed: Bool { errorMessage == nil }
}

// MARK: - Benchmark Result

struct BenchmarkResult: Codable, Sendable, Identifiable {
    let id: UUID
    let timestamp: Date
    let category: BenchmarkCategory
    let scenario: BenchmarkScenario
    let modelInfo: ComponentModelInfo
    let metrics: BenchmarkMetrics

    init(
        category: BenchmarkCategory,
        scenario: BenchmarkScenario,
        modelInfo: ComponentModelInfo,
        metrics: BenchmarkMetrics
    ) {
        self.id = UUID()
        self.timestamp = Date()
        self.category = category
        self.scenario = scenario
        self.modelInfo = modelInfo
        self.metrics = metrics
    }
}

// MARK: - Benchmark Run

struct BenchmarkRun: Codable, Sendable, Identifiable {
    let id: UUID
    let startedAt: Date
    var completedAt: Date?
    var results: [BenchmarkResult]
    var status: BenchmarkRunStatus
    let deviceInfo: BenchmarkDeviceInfo

    var duration: TimeInterval? {
        guard let completedAt else { return nil }
        return completedAt.timeIntervalSince(startedAt)
    }

    init(deviceInfo: BenchmarkDeviceInfo) {
        self.id = UUID()
        self.startedAt = Date()
        self.results = []
        self.status = .running
        self.deviceInfo = deviceInfo
    }
}

// MARK: - Benchmark Run Output

/// Returned by `BenchmarkRunner.runBenchmarks` so callers get both results
/// and the categories that were skipped (no downloaded models) in one shot,
/// eliminating the need for a redundant preflight call.
struct BenchmarkRunOutput: Sendable {
    let results: [BenchmarkResult]
    let skippedCategories: [BenchmarkCategory]
}

// MARK: - Progress Update

struct BenchmarkProgressUpdate: Sendable {
    let completedCount: Int
    let totalCount: Int
    let currentScenario: String
    let currentModel: String

    var progress: Double {
        guard totalCount > 0 else { return 0 }
        return Double(completedCount) / Double(totalCount)
    }
}

File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/benchmarks/models/BenchmarkTypes.kt
package com.runanywhere.runanywhereai.presentation.benchmarks.models

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ChatBubble
import androidx.compose.material.icons.filled.GraphicEq
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material.icons.automirrored.filled.VolumeUp
import androidx.compose.ui.graphics.vector.ImageVector
import com.runanywhere.sdk.public.extensions.Models.ModelCategory
import com.runanywhere.sdk.public.extensions.Models.ModelInfo
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID

// -- Benchmark Category --

@Serializable
enum class BenchmarkCategory(val value: String) {
    @SerialName("llm") LLM("llm"),
    @SerialName("stt") STT("stt"),
    @SerialName("tts") TTS("tts"),
    @SerialName("vlm") VLM("vlm"),
    ;

    val displayName: String
        get() = when (this) {
            LLM -> "LLM"
            STT -> "STT"
            TTS -> "TTS"
            VLM -> "VLM"
        }

    val icon: ImageVector
        get() = when (this) {
            LLM -> Icons.Filled.ChatBubble
            STT -> Icons.Filled.GraphicEq
            TTS -> Icons.AutoMirrored.Filled.VolumeUp
            VLM -> Icons.Filled.Visibility
        }

    val modelCategory: ModelCategory
        get() = when (this) {
            LLM -> ModelCategory.LANGUAGE
            STT -> ModelCategory.SPEECH_RECOGNITION
            TTS -> ModelCategory.SPEECH_SYNTHESIS
            VLM -> ModelCategory.MULTIMODAL
        }
}

// -- Benchmark Run Status --

@Serializable
enum class BenchmarkRunStatus(val value: String) {
    @SerialName("running") RUNNING("running"),
    @SerialName("completed") COMPLETED("completed"),
    @SerialName("failed") FAILED("failed"),
    @SerialName("cancelled") CANCELLED("cancelled"),
}

// -- Benchmark Scenario --

@Serializable
data class BenchmarkScenario(
    val name: String,
    val category: BenchmarkCategory,
) {
    val id: String get() = "${category.value}_$name"
}

// -- Component Model Info (snapshot of ModelInfo for persistence) --

@Serializable
data class ComponentModelInfo(
    val id: String,
    val name: String,
    val framework: String,
    val category: String,
) {
    companion object {
        fun from(model: ModelInfo): ComponentModelInfo = ComponentModelInfo(
            id = model.id,
            name = model.name,
            framework = model.framework.displayName,
            category = model.category.value,
        )
    }
}

// -- Device Info (snapshot for persistence) --

@Serializable
data class BenchmarkDeviceInfo(
    val modelName: String,
    val chipName: String,
    val totalMemoryBytes: Long,
    val availableMemoryBytes: Long,
    val osVersion: String,
)

// -- Benchmark Metrics --

@Serializable
data class BenchmarkMetrics(
    // Common
    val endToEndLatencyMs: Double = 0.0,
    val loadTimeMs: Double = 0.0,
    val warmupTimeMs: Double = 0.0,
    val memoryDeltaBytes: Long = 0,

    // LLM-specific
    val ttftMs: Double? = null,
    val tokensPerSecond: Double? = null,
    val inputTokens: Int? = null,
    val outputTokens: Int? = null,

    // STT-specific
    val audioLengthSeconds: Double? = null,
    val realTimeFactor: Double? = null,

    // TTS-specific
    val audioDurationSeconds: Double? = null,
    val charactersProcessed: Int? = null,

    // VLM-specific
    val promptTokens: Int? = null,
    val completionTokens: Int? = null,

    // Error info
    val errorMessage: String? = null,
) {
    val didSucceed: Boolean get() = errorMessage == null
}

// -- Benchmark Result --

@Serializable
data class BenchmarkResult(
    val id: String = UUID.randomUUID().toString(),
    val timestamp: Long = System.currentTimeMillis(),
    val category: BenchmarkCategory,
    val scenario: BenchmarkScenario,
    val modelInfo: ComponentModelInfo,
    val metrics: BenchmarkMetrics,
)

// -- Benchmark Run --

@Serializable
data class BenchmarkRun(
    val id: String = UUID.randomUUID().toString(),
    val startedAt: Long = System.currentTimeMillis(),
    val completedAt: Long? = null,
    val results: List<BenchmarkResult> = emptyList(),
    val status: BenchmarkRunStatus = BenchmarkRunStatus.RUNNING,
    val deviceInfo: BenchmarkDeviceInfo,
) {
    val durationSeconds: Double?
        get() {
            val completed = completedAt ?: return null
            return (completed - startedAt) / 1000.0
        }
}

// -- Progress Update --

data class BenchmarkProgressUpdate(
    val completedCount: Int,
    val totalCount: Int,
    val currentScenario: String,
    val currentModel: String,
) {
    val progress: Float
        get() = if (totalCount > 0) completedCount.toFloat() / totalCount.toFloat() else 0f
}
